File: src/main/java/org/cubexmc/railway/placeholder/RailwayPlaceholders.java
package org.cubexmc.railway.placeholder;

import org.bukkit.OfflinePlayer;
import org.cubexmc.railway.Railway;
import org.cubexmc.railway.model.Line;
import org.cubexmc.railway.model.Stop;
import org.cubexmc.railway.service.LineService;
import org.cubexmc.railway.train.TrainInstance;

import me.clip.placeholderapi.expansion.PlaceholderExpansion;

import java.util.List;

public class RailwayPlaceholders extends PlaceholderExpansion {

    private final Railway plugin;

    public RailwayPlaceholders(Railway plugin) {
        this.plugin = plugin;
    }

    @Override
    public String getAuthor() {
        return plugin.getDescription().getAuthors().isEmpty() ? "railway" : String.join(", ", plugin.getDescription().getAuthors());
    }

    @Override
    public String getIdentifier() {
        return "railway";
    }

    @Override
    public String getVersion() {
        return plugin.getDescription().getVersion();
    }

    @Override
    public boolean persist() {
        return true;
    }

    @Override
    public boolean canRegister() {
        return true;
    }

    @Override
    public String onRequest(OfflinePlayer player, String params) {
        if (params == null || params.isEmpty()) return "";
        String[] parts = params.split("_");
        if (parts.length == 0) return "";
        String type = parts[0].toLowerCase();

        if ("eta".equals(type)) {
            // %railway_eta_line_l1_stop_s3%
            String lineId = findValue(parts, "line");
            String stopId = findValue(parts, "stop");
            if (lineId == null || stopId == null) return "";
            int etaSec = estimateEtaSeconds(lineId, stopId);
            if (etaSec < 0) return "--:--";
            return formatSeconds(etaSec);
        }

        if ("next".equals(type) || "nextstop".equals(type)) {
            // %railway_next_line_l1_from_s3%
            String lineId = findValue(parts, "line");
            String fromId = findValue(parts, "from");
            if (lineId == null || fromId == null) return "";
            return nextStopName(lineId, fromId);
        }

        return "";
    }

    private String findValue(String[] parts, String key) {
        for (int i = 0; i < parts.length - 1; i++) {
            if (key.equalsIgnoreCase(parts[i])) {
                return parts[i + 1];
            }
        }
        // Also support simple pattern eta_l1_s3
        if ("line".equals(key) && parts.length >= 3) return parts[1];
        if ("stop".equals(key) && parts.length >= 3) return parts[2];
        if ("from".equals(key) && parts.length >= 3) return parts[2];
        return null;
    }

    private int estimateEtaSeconds(String lineId, String stopId) {
        LineService service = plugin.getLineServiceManager().getService(lineId);
        if (service == null) {
            return -1;
        }
        // Prefer active trains' ETA
        int best = Integer.MAX_VALUE;
        long now = org.bukkit.Bukkit.getCurrentTick();
        for (TrainInstance t : service.getActiveTrains()) {
            double eta = t.estimateEtaSecondsToStop(stopId, now, plugin.getTravelTimeEstimator());
            if (eta != Double.POSITIVE_INFINITY) {
                int sec = (int) Math.round(eta);
                if (sec < best) best = sec;
            }
        }
        if (best != Integer.MAX_VALUE) return best;
        // Fallback to headway-based estimate
        return plugin.getLineServiceManager().estimateNextEtaSeconds(lineId, stopId);
    }

    private String nextStopName(String lineId, String fromId) {
        Line line = plugin.getLineManager().getLine(lineId);
        if (line == null) return "";
        List<String> stops = line.getOrderedStopIds();
        if (stops == null || stops.isEmpty()) return "";
        int idx = stops.indexOf(fromId);
        if (idx < 0) return "";
        boolean loop = plugin.getLineServiceManager().getService(lineId) != null
                && plugin.getLineServiceManager().getService(lineId).isLoopLine();
        int nextIdx = idx + 1;
        if (nextIdx >= stops.size()) {
            if (loop) nextIdx = 0; else return "";
        }
        String nextId = stops.get(nextIdx);
        Stop s = plugin.getStopManager().getStop(nextId);
        return (s != null && s.getName() != null && !s.getName().isEmpty()) ? s.getName() : nextId;
    }

    private String formatSeconds(int seconds) {
        if (seconds < 0) seconds = 0;
        int m = seconds / 60;
        int s = seconds % 60;
        return String.format("%d:%02d", m, s);
    }
}

File: src/main/java/org/cubexmc/metro/placeholder/RailwayPlaceholders.kt
package org.cubexmc.metro.placeholder

import java.util.Locale
import org.bukkit.OfflinePlayer
import org.cubexmc.metro.Metro
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.service.LineService
import org.cubexmc.metro.train.TrainInstance
import org.cubexmc.metro.util.SchedulerUtil

class RailwayPlaceholders(
    private val plugin: Metro,
) : NullablePlaceholderExpansion() {
    override fun getAuthor(): String {
        val authors = plugin.description.authors
        return if (authors.isEmpty()) "railway" else authors.joinToString(", ")
    }

    override fun getIdentifier(): String = "railway"

    override fun getVersion(): String = plugin.description.version

    override fun persist(): Boolean = true

    override fun canRegister(): Boolean = true

    override fun onRequest(player: OfflinePlayer?, params: String?): String {
        if (params.isNullOrEmpty()) {
            return ""
        }
        val parts = params.split('_').dropLastWhile { it.isEmpty() }.toTypedArray()
        if (parts.isEmpty()) {
            return ""
        }
        return when (parts[0].lowercase(Locale.getDefault())) {
            "eta" -> {
                // %railway_eta_line_l1_stop_s3%
                val lineId = findValue(parts, "line")
                val stopId = findValue(parts, "stop")
                if (lineId == null || stopId == null) {
                    ""
                } else {
                    val etaSeconds = estimateEtaSeconds(lineId, stopId)
                    if (etaSeconds < 0) "--:--" else formatSeconds(etaSeconds)
                }
            }

            "next", "nextstop" -> {
                // %railway_next_line_l1_from_s3%
                val lineId = findValue(parts, "line")
                val fromId = findValue(parts, "from")
                if (lineId == null || fromId == null) "" else nextStopName(lineId, fromId)
            }

            else -> ""
        }
    }

    private fun findValue(parts: Array<String>, key: String): String? {
        for (index in 0 until parts.size - 1) {
            if (key.equals(parts[index], ignoreCase = true)) {
                return parts[index + 1]
            }
        }
        // Also support simple pattern eta_l1_s3
        if (key == "line" && parts.size >= 3) {
            return parts[1]
        }
        if (key == "stop" && parts.size >= 3) {
            return parts[2]
        }
        if (key == "from" && parts.size >= 3) {
            return parts[2]
        }
        return null
    }

    private fun estimateEtaSeconds(lineId: String, stopId: String): Int {
        val service: LineService = plugin.lineServiceManager.getService(lineId) ?: return -1
        var best = Int.MAX_VALUE
        val now = SchedulerUtil.getCurrentTick()
        for (train: TrainInstance in service.activeTrains) {
            val eta = train.estimateEtaSecondsToStop(stopId, now, plugin.travelTimeEstimator)
            if (eta != Double.POSITIVE_INFINITY) {
                val seconds = Math.round(eta).toInt()
                if (seconds < best) {
                    best = seconds
                }
            }
        }
        if (best != Int.MAX_VALUE) {
            return best
        }
        return plugin.lineServiceManager.estimateNextEtaSeconds(lineId, stopId)
    }

    private fun nextStopName(lineId: String, fromId: String): String {
        val line: Line = plugin.lineManager.getLine(lineId) ?: return ""
        val stops: List<String>? = line.orderedStopIds
        if (stops.isNullOrEmpty()) {
            return ""
        }
        val index = stops.indexOf(fromId)
        if (index < 0) {
            return ""
        }
        val service = plugin.lineServiceManager.getService(lineId)
        val loop = service != null && service.isLoopLine
        var nextIndex = index + 1
        if (nextIndex >= stops.size) {
            if (loop) {
                nextIndex = 0
            } else {
                return ""
            }
        }
        val nextId = stops[nextIndex]
        val stop: Stop? = plugin.stopManager.getStop(nextId)
        val stopName: String? = stop?.name
        return if (!stopName.isNullOrEmpty()) stopName else nextId
    }

    private fun formatSeconds(seconds: Int): String {
        val clampedSeconds = maxOf(0, seconds)
        val minutes = clampedSeconds / 60
        val remainingSeconds = clampedSeconds % 60
        return String.format("%d:%02d", minutes, remainingSeconds)
    }
}
